core/application/bloodDonationWorkflow/DonorSearchService.ts

```diff
@@ -0,0 +1,182 @@
+import {
+  DonationStatus,
+  DonorSearchDTO
+} from '../../../commons/dto/DonationDTO'
+import Repository from '../models/policies/repositories/Repository'
+import { getGeohashNthNeighbors } from '../utils/geohash'
+import {
+  QueryConditionOperator,
+  QueryInput
+} from '../models/policies/repositories/QueryTypes'
+import {
+  DonorRoutingAttributes,
+  StepFunctionInput
+} from './Types'
+import { StepFunctionModel } from '../models/stepFunctions/StepFunctionModel'
+import { DONOR_SEARCH_PK_PREFIX } from '../models/dbModels/DonorSearchModel'
+import { AcceptedDonationFields } from '../models/dbModels/AcceptDonationModel'
+import { LocationDTO, UserDetailsDTO } from '../../../commons/dto/UserDTO'
+import { getBloodRequestMessage } from './BloodDonationMessages'
+import LocationModel from '../models/dbModels/LocationModel'
+
+export class DonorSearchService {
+  async routeDonorRequest(
+    donorRoutingAttributes: DonorRoutingAttributes,
+    sourceQueueArn: string,
+    userProfile: UserDetailsDTO,
+    donorSearchRepository: Repository<DonorSearchDTO>,
+    stepFunctionModel: StepFunctionModel
+  ): Promise<void> {
+    const { seekerId, requestPostId, createdAt } = donorRoutingAttributes
+
+    const donorSearchRecord = await donorSearchRepository.getItem(
+      `${DONOR_SEARCH_PK_PREFIX}#${seekerId}`,
+      `${DONOR_SEARCH_PK_PREFIX}#${createdAt}#${requestPostId}`
+    )
+
+    if (donorSearchRecord === null) {
+      await donorSearchRepository.create({
+        id: donorRoutingAttributes.requestPostId,
+        ...donorRoutingAttributes,
+        status: DonationStatus.PENDING,
+        retryCount: 0
+      })
+    }
+
+    if (donorSearchRecord !== null && donorSearchRecord.status === DonationStatus.COMPLETED) {
```

Comment:
I am not sure I follow this logic here, 
if DonationStatus is COMPLETED, we set it to be PENDING?
that is a bit strange
Semantically it would go the opposite way no?

Reply:
also do you not want to do something like this instead
```
const isValidQueue = sourceQueueArn === process.env.DONOR_SEARCH_QUEUE_ARN // not sure why this check is important
const hasDonationCompleted = donorSearchRecord !== null && donorSearchRecord.status === DonationStatus.COMPLETED
const notEnoughBloodDiscovered = donorRoutingAttributes.bloodQuantity > donorSearchRecord.bloodQuantity // not really sure what this is doing to be honest
if (isValidQueue && hasDonationCompleted && notEnoughBloodDiscovered) {
    await donorSearchRepository.update({
        ...donorRoutingAttributes,
        id: requestPostId,
        status: DonationStatus.PENDING,
        retryCount: 0
    })
}

```

Reply:
if its coming from the DONOR_SEARCH_QUEUE_ARN, its a donation update request. that's why I'm setting it as pending to start again.

and update the code with your code

Reply:
then lets use that as the variable name
```
const isDonationUpdateRequest = sourceQueueArn === process.env.DONOR_SEARCH_QUEUE_ARN
```

Reply:
Updated in 4aa41965d2a6ddef1f79ae833580099c7690bdd6